cartographer_ros/cartographer_ros/metrics/internal/histogram.h

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2018 The Cartographer Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CARTOGRAPHER_ROS_METRICS_HISTOGRAM_H
+#define CARTOGRAPHER_ROS_METRICS_HISTOGRAM_H
+
+#include <algorithm>
+#include <map>
+#include <numeric>
+#include <vector>
+
+#include "cartographer/common/mutex.h"
+#include "cartographer/metrics/histogram.h"
+#include "cartographer_ros/metrics/internal/counter.h"
+#include "cartographer_ros_msgs/Metric.h"
+#include "glog/logging.h"
+
+namespace cartographer_ros {
+namespace metrics {
+
+constexpr double kInfiniteBoundary = std::numeric_limits<double>::infinity();
+
+using BucketBoundaries = ::cartographer::metrics::Histogram::BucketBoundaries;
+
+class Histogram : public ::cartographer::metrics::Histogram {
+ public:
+  Histogram(const std::map<std::string, std::string>& labels,
+            const BucketBoundaries& bucket_boundaries)
+      : labels_(labels),
+        bucket_boundaries_(bucket_boundaries),
+        bucket_counts_(bucket_boundaries.size() + 1) {
+    ::cartographer::common::MutexLocker lock(&mutex_);
+    CHECK(std::is_sorted(std::begin(bucket_boundaries_),
+                         std::end(bucket_boundaries_)));
+  }
+
+  Histogram(const BucketBoundaries& bucket_boundaries)
+      : Histogram({}, bucket_boundaries) {}
+
+  void Observe(double value) override {
+    ::cartographer::common::MutexLocker lock(&mutex_);
```

Comment:
add todo to check lock contention here and potentially re-implement with a vector of atomics.